### PR TITLE
[file_selector] Add file group to save return value

### PR DESCRIPTION
**Pull request by stuartmorgan**
This deprecates `getSavePath`, which returned a target path string, in favor of a new `getSaveLocation`, which returns an object containing both a path and, optionally, a selected file group. This allows clients to use the selected group when deciding what path to use when saving (see discussion in linked issue).

This includes an implementation for Windows. It will also apply to Linux, and I've verified that the structure works, but it's not included here because it requires some non-trivial refactoring in the Linux implementation (we can't get the current index, only the current filter object pointer, which means we need to pass more data around between the various functions to map back to an index... and it's GObject so making internal data utility classes is fiddly.) For now Linux just always returns a null group, and we can add it later.

Most of https://github.com/flutter/flutter/issues/107093

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/file_selector/file_selector/test/file_selector_test.dart [hunk 1]
```diff
@@ -374,12 +448,35 @@ class FakeFileSelector extends Fake
...
+  }) async {
+    return (await getSaveLocation(
+      acceptedTypeGroups: acceptedTypeGroups,
+      options: SaveDialogOptions(
+        initialDirectory: initialDirectory,
+        suggestedName: suggestedName,
+        confirmButtonText: confirmButtonText,
+      ),
+    ))
+        ?.path;
```
Comment (cbracken) [inline packages/file_selector/file_selector/test/file_selector_test.dart:460]: Was this intentional (or produced with dartfmt)?

File: packages/file_selector/file_selector_platform_interface/test/file_selector_platform_interface_test.dart [hunk 2]
```diff
@@ -29,6 +29,30 @@ void main() {
...
+    expect(result?.path, OldFileSelectorPlatformImplementation.savePath);
+    expect(result?.activeFilter, null);
+  });
 }
 
 class ExtendsFileSelectorPlatform extends FileSelectorPlatform {}
+
+class OldFileSelectorPlatformImplementation extends FileSelectorPlatform {
+  static const String savePath = '/a/path';
+// Only implement the deprecated getSavePath.
```
Comment (cbracken) [inline packages/file_selector/file_selector_platform_interface/test/file_selector_platform_interface_test.dart:48]: pedantic nit: indent

File: packages/file_selector/file_selector_windows/example/lib/save_text_page.dart [hunk 3]
```diff
@@ -17,19 +18,37 @@ class SaveTextPage extends StatelessWidget {
...
+    if (result == null) {
       return;
     }
+    String path = result.path;
+    // Append an extension based on the selected type group if the user didn't
+    // include one.
+    if (!path.split(Platform.pathSeparator).last.contains('.')) {
+      final XTypeGroup? activeGroup = result.activeFilter;
+      if (activeGroup != null) {
+        path = '$path.${activeGroup.extensions!.first}';
```
Comment (cbracken) [inline packages/file_selector/file_selector_windows/example/lib/save_text_page.dart:45]: Is the extensions list always guaranteed to contain at least one so long as there's an activeFilter? (That makes logical sense but it's a bummer we need to know it here)

File: packages/file_selector/file_selector_windows/windows/file_selector_plugin.cpp [hunk 4]
```diff
@@ -190,7 +190,14 @@ class DialogWrapper {
       }
       files.push_back(EncodableValue(GetPathForShellItem(shell_item)));
     }
-    return files;
+    FileDialogResult result(files, nullptr);
+    UINT file_type_index;
+    if (SUCCEEDED(dialog_controller_->GetFileTypeIndex(&file_type_index)) &&
+        file_type_index > 0) {
```
Comment (cbracken) [inline packages/file_selector/file_selector_windows/windows/file_selector_plugin.cpp:196]: 😢 If this were pascal...

File: packages/file_selector/file_selector_windows/windows/test/file_selector_plugin_test.cpp [hunk 5]
```diff
@@ -341,17 +322,78 @@ TEST(FileSelectorPlugin, TestSaveWithArguments) {
...
+  ErrorOr<FileDialogResult> result = plugin.ShowSaveDialog(
+      SelectionOptions(/* allow multiple = */ false,
+                       /* select folders = */ false, EncodableList()),
       &initial_directory, &suggested_name, &confirm_button);
 
   EXPECT_TRUE(shown);
   ASSERT_FALSE(result.has_error());
-  const EncodableList& paths = result.value();
+  const EncodableList& paths = result.value().paths();
+  EXPECT_EQ(paths.size(), 1);
```
Comment (cbracken) [inline packages/file_selector/file_selector_windows/windows/test/file_selector_plugin_test.cpp:333]: Given the next line, arguably `ASSERT_EQ`.

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

File: packages/file_selector/file_selector_windows/example/lib/save_text_page.dart (hunk 3 above)
Comment (stuartmorgan) [reply packages/file_selector/file_selector_windows/example/lib/save_text_page.dart:45]: This is dependent on the code 15-20 lines up where we define the groups. I'll add a comment explaining that so it's clear to a reader why it's safe.

**Review by ditman — APPROVED** [review]
Wow, what a big PR for a relatively small fix/feature!

The PR LGTM (I skipped .g.dart and .mock.dart files). Web bits look good (since they're pretty much a Noop). cpp stuff looks... almost intelligible :P

Thanks for fixing this!

File: packages/file_selector/file_selector/test/file_selector_test.dart [hunk 6]
```diff
@@ -374,12 +448,35 @@ class FakeFileSelector extends Fake
     String? initialDirectory,
     String? suggestedName,
     String? confirmButtonText,
+  }) async {
+    return (await getSaveLocation(
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/file_selector/file_selector/test/file_selector_test.dart]: Done.

File: packages/file_selector/file_selector/test/file_selector_test.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/file_selector/file_selector/test/file_selector_test.dart]: `dartfmt` :( But @ditman's suggestion fixed it.

File: packages/file_selector/file_selector_platform_interface/lib/src/types/file_save_location_result.dart [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+import 'package:flutter/foundation.dart' show immutable;
+
+import 'x_type_group.dart';
+
+export 'x_type_group.dart';
+
+/// The response from a save dialog.
+@immutable
+class FileSaveLocationResult {
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/file_selector/file_selector_platform_interface/lib/src/types/file_save_location_result.dart]: That's a good point, I'll change it.

File: packages/file_selector/file_selector_platform_interface/test/file_selector_platform_interface_test.dart (hunk 2 above)
Comment (stuartmorgan) [reply packages/file_selector/file_selector_platform_interface/test/file_selector_platform_interface_test.dart]: Fixed. I've never figured out when `dart format` *doesn't* fix comment indentation.

File: packages/file_selector/file_selector_windows/windows/test/file_selector_plugin_test.cpp (hunk 5 above)
Comment (stuartmorgan) [reply packages/file_selector/file_selector_windows/windows/test/file_selector_plugin_test.cpp]: Fixed everywhere.

**Comment by stuartmorgan** [comment]
I'll split out the first sub-PR shortly.

**Pull request merged**